File: bench/matmul/matmul.ts
// TypeScript peer of bench/matmul.vader. An array of arrays, and the row is
// hoisted the same way, so the comparison is of the indexing and not of the
// loop shape. Values stay well inside an f64 mantissa.

const SIZE = 260;

function build(modulus: number): number[][] {
  const out: number[][] = [];
  for (let row = 0; row < SIZE; row++) {
    const line: number[] = [];
    for (let column = 0; column < SIZE; column++) {
      line.push((row + column) % modulus);
    }
    out.push(line);
  }
  return out;
}

const left = build(7);
const right = build(5);
let total = 0;
for (let row = 0; row < SIZE; row++) {
  const leftRow = left[row]!;
  for (let column = 0; column < SIZE; column++) {
    let cell = 0;
    for (let k = 0; k < SIZE; k++) {
      cell += leftRow[k]! * right[k]![column]!;
    }
    total += cell;
  }
}
console.log(`matmul size=${SIZE} total=${total}`);

File: bench/matmul/matmul.go
// Go peer of bench/matmul.vader. A [][]int64 is a slice of slices, so the row
// lookup is the same shape as Vader's array of arrays.

package main

import "fmt"

const size = 260

func build(modulus int64) [][]int64 {
	out := make([][]int64, 0, size)
	for row := 0; row < size; row++ {
		line := make([]int64, 0, size)
		for column := 0; column < size; column++ {
			line = append(line, int64(row+column)%modulus)
		}
		out = append(out, line)
	}
	return out
}

func main() {
	left := build(7)
	right := build(5)
	var total int64
	for row := 0; row < size; row++ {
		leftRow := left[row]
		for column := 0; column < size; column++ {
			var cell int64
			for k := 0; k < size; k++ {
				cell += leftRow[k] * right[k][column]
			}
			total += cell
		}
	}
	fmt.Printf("matmul size=%d total=%d\n", size, total)
}
